BUG: Cannot map MultiIndex to more levels

GPla opened the pull request:
- [x] closes #24800 
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

GPla committed:
BUG: Cannot map MultiIndex to more levels

GPla committed:
BUG: Cannot map MultiIndex to more levels 846dc78

fixed formatting

GPla committed:
BUG: Cannot map MultiIndex to more levels

fixed issue number in doc

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -6364,7 +6364,7 @@ def map(self, mapper, na_action=None):
         # we can return a MultiIndex
         if new_values.size and isinstance(new_values[0], tuple):
-            if isinstance(self, MultiIndex):
+            if isinstance(self, MultiIndex) and len(new_values[0]) == self.nlevels:
```
what happens if its a Multiindex bu tthis is not true? do we have a test case for that

GPla replied:
The checks that happen here are only to determine the `names` of the new index. 

```python
if isinstance(self, MultiIndex) and len(new_values[0]) == self.nlevels:
     names = self.names
 elif self.name:
    names = [self.name] * len(new_values[0])
else:
  names = None
```

If the old index is a `MultiIndex` then we reuse  `names` for the new index. However, if the number of levels changed then `len(self.names) != len(len(new_values[0])` and an error will be raised. With the change, `names` is not preserved and reset as seen in the snippet. The last of the three tests that I added checks for the new case.

simonjayhawkins commented on doc/source/whatsnew/v1.5.0.rst:
```diff
@@ -802,6 +802,7 @@ MultiIndex
 - Bug in :meth:`Series.to_numpy` where multiindexed Series could not be converted to numpy arrays when an ``na_value`` was supplied (:issue:`45774`)
 - Bug in :class:`MultiIndex.equals` not commutative when only one side has extension array dtype (:issue:`46026`)
 - Bug in :meth:`MultiIndex.from_tuples` cannot construct Index of empty tuples (:issue:`45608`)
+- Bug in :meth:`Index.map` cannot map an existing :class:`MultiIndex` to more levels (:issue:`47173`)
```
```suggestion
- Bug in :meth:`Index.map` cannot map an existing :class:`MultiIndex` to more levels (:issue:`24800`)
```

GPla committed:
BUG: Cannot map MultiIndex to more levels

update changelog

Co-authored-by: Simon Hawkins <simonjayhawkins@gmail.com>

GPla committed:
BUG: Cannot map MultiIndex to more levels

removed name preservation in Index.map

GPla commented:
@simonjayhawkins I updated the code to reflect [#24800 (comment)](https://github.com/pandas-dev/pandas/issues/24800#issuecomment-455069121).

Furthermore, I removed the name preservation completely. Keeping the other check.

```python
if self.name:
    names = [self.name] * len(new_values[0])
else:
    names = None
```

would set `names` the first time we switch from an `Index` to a `MultiIndex` but would be lost in successive `map` operations since `MultiIndex` does not set/have a `name` property. Thus, I removed this check as well.

simonjayhawkins commented:
> EDIT: There are 5 failing tests which expect a name preservation from `MultiIndex` to `MultiIndex` and one test which expects a name preservation from `Index` to `MultiIndex`. How should this be resolved?

we may need to deprecate the existing behavior. cc @jorisvandenbossche @toobaz  (from #24800 discussion)
